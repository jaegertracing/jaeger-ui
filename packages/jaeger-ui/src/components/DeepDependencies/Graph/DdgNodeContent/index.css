/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
 * Because .plexus-Digraph--MeasurableHtmlNode is `position: absolute`, adding a z-index to DdgNodeContent
 * would have no effect on layering between different nodes.
 */
.plexus-Digraph--MeasurableHtmlNode {
  /* transition-delay must equal .DdgNodeContent--actionsWrapper transition delay plus duration */
  transition-delay: 250ms;
  transition-property: z-index;
  z-index: 0;
}

.plexus-Digraph--MeasurableHtmlNode:hover {
  transition-delay: 0s;
  z-index: 1;
}

.DdgNodeContent--core {
  background: #eee;
  border-radius: 100%;
  box-shadow: 0 1px 2px 1px rgba(0, 0, 0, 0.15);
  cursor: default;
  display: flex;
}

.DdgNodeContent--core.is-decorated {
  cursor: pointer;
}

.DdgNodeContent--core.is-missingDecoration {
  background-color: #e58c33;
}

.DdgNodeContent--core.is-positioned {
  bottom: 1px;
  left: 1px;
  position: absolute;
  right: 1px;
  top: 1px;
}

.DdgNodeContent--labelWrapper {
  margin-right: auto;
  margin-left: auto;
}

.DdgNodeContent--label {
  margin: 0;
  margin-right: auto;
  margin-left: auto;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.DdgNodeContent--actionsWrapper {
  background: #fff;
  border-radius: 3px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  bottom: 100%;
  box-shadow: 0 0px 4px 1px rgba(0, 0, 0, 0.1);
  left: 1em;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  /* transition delay plus duration must equal .plexus-Digraph--MeasurableHtmlNode transition-delay  */
  transition-delay: 150ms;
  transition-duration: 0.1s;
  transition-property: opacity;
}

.DdgNodeContent:hover > .DdgNodeContent--actionsWrapper {
  opacity: 1;
  pointer-events: all;
  transition-delay: 0s;
}

.DdgNodeContent--actionsItem {
  align-items: center;
  display: flex;
  padding: 0.5em;
}

.DdgNodeContent--actionsItem:not(:hover) {
  color: inherit;
}

.DdgNodeContent--actionsItemIconWrapper {
  flex: none;
  height: 16px;
  width: 16px;
}

.DdgNodeContent--actionsItemIconWrapper > * {
  height: 16px;
  position: absolute;
  width: 16px;
}

.DdgNodeContent--actionsItemIconWrapper > * > .ant-checkbox {
  vertical-align: unset;
  top: unset;
}

.DdgNodeContent--actionsItemText {
  font-size: 0.9em;
  line-height: normal;
  margin-left: 0.5em;
}
