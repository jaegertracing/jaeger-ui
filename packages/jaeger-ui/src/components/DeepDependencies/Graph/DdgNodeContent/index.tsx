// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Checkbox, Popover } from 'antd';
import cx from 'classnames';
import { TLayoutVertex } from '@jaegertracing/plexus/lib/types';
import IoAndroidLocate from 'react-icons/lib/io/android-locate';
import MdVisibilityOff from 'react-icons/lib/md/visibility-off';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import calcPositioning from './calc-positioning';
import {
  MAX_LENGTH,
  MAX_LINKED_TRACES,
  MIN_LENGTH,
  OP_PADDING_TOP,
  PARAM_NAME_LENGTH,
  PROGRESS_BAR_STROKE_WIDTH,
  RADIUS,
  WORD_RX,
} from './constants';
import { setFocusIcon } from './node-icons';
import { trackSetFocus, trackViewTraces, trackVertexSetOperation } from '../../index.track';
import { getUrl } from '../../url';
import BreakableText from '../../../common/BreakableText';
import FilteredList from '../../../common/FilteredList';
import NewWindowIcon from '../../../common/NewWindowIcon';
import { getUrl as getSearchUrl } from '../../../SearchTracePage/url';
import padActions from '../../../../actions/path-agnostic-decorations';
import {
  ECheckedStatus,
  EDdgDensity,
  EDirection,
  EViewModifier,
  TDdgVertex,
  PathElem,
} from '../../../../model/ddg/types';
import extractDecorationFromState, {
  TDecorationFromState,
} from '../../../../model/path-agnostic-decorations';
import { ReduxState } from '../../../../types/index';

import './index.css';

type TDispatchProps = {
  getDecoration: (id: string, svc: string, op?: string) => void;
};

type TProps = TDispatchProps &
  TDecorationFromState & {
    focalNodeUrl: string | null;
    focusPathsThroughVertex: (vertexKey: string) => void;
    getGenerationVisibility: (vertexKey: string, direction: EDirection) => ECheckedStatus | null;
    getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
    hideVertex: (vertexKey: string) => void;
    isFocalNode: boolean;
    isPositioned: boolean;
    operation: string | string[] | null;
    selectVertex: (selectedVertex: TDdgVertex) => void;
    service: string;
    setOperation: (operation: string) => void;
    setViewModifier: (visIndices: number[], viewModifier: EViewModifier, isEnabled: boolean) => void;
    updateGenerationVisibility: (vertexKey: string, direction: EDirection) => void;
    vertex: TDdgVertex;
    vertexKey: string;
  };

type TState = {
  childrenVisibility?: ECheckedStatus | null;
  parentVisibility?: ECheckedStatus | null;
};

export function getNodeRenderer({
  baseUrl,
  density,
  extraUrlArgs,
  focusPathsThroughVertex,
  getGenerationVisibility,
  getVisiblePathElems,
  hideVertex,
  selectVertex,
  setOperation,
  setViewModifier,
  updateGenerationVisibility,
}: {
  baseUrl: string;
  density: EDdgDensity;
  extraUrlArgs?: { [key: string]: unknown };
  focusPathsThroughVertex: (vertexKey: string) => void;
  getGenerationVisibility: (vertexKey: string, direction: EDirection) => ECheckedStatus | null;
  getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
  hideVertex: (vertexKey: string) => void;
  selectVertex: (selectedVertex: TDdgVertex) => void;
  setOperation: (operation: string) => void;
  setViewModifier: (visIndices: number[], viewModifier: EViewModifier, isEnabled: boolean) => void;
  updateGenerationVisibility: (vertexKey: string, direction: EDirection) => void;
}) {
  return function renderNode(vertex: TDdgVertex, _: unknown, lv: TLayoutVertex<any> | null) {
    const { isFocalNode, key, operation, service } = vertex;
    return (
      <DdgNodeContent
        focalNodeUrl={isFocalNode ? null : getUrl({ density, operation, service, ...extraUrlArgs }, baseUrl)}
        focusPathsThroughVertex={focusPathsThroughVertex}
        getGenerationVisibility={getGenerationVisibility}
        getVisiblePathElems={getVisiblePathElems}
        hideVertex={hideVertex}
        isFocalNode={isFocalNode}
        isPositioned={Boolean(lv)}
        operation={operation}
        selectVertex={selectVertex}
        setOperation={setOperation}
        setViewModifier={setViewModifier}
        service={service}
        updateGenerationVisibility={updateGenerationVisibility}
        vertex={vertex}
        vertexKey={key}
      />
    );
  };
}

export function measureNode() {
  const diameter = 2 * (RADIUS + 1);

  return {
    height: diameter,
    width: diameter,
  };
}

export class UnconnectedDdgNodeContent extends React.PureComponent<TProps, TState> {
  state: TState = {};
  hoveredIndices: Set<number> = new Set();

  constructor(props: TProps) {
    super(props);

    this.getDecoration();
  }

  componentDidUpdate(prevProps: TProps) {
    if (prevProps.decorationID !== this.props.decorationID) this.getDecoration();
  }

  componentWillUnmount() {
    if (this.hoveredIndices.size) {
      this.props.setViewModifier(Array.from(this.hoveredIndices), EViewModifier.Hovered, false);
      this.hoveredIndices.clear();
    }
  }

  private getDecoration() {
    const { decorationID, getDecoration, operation, service } = this.props;

    if (decorationID) {
      getDecoration(decorationID, service, typeof operation === 'string' ? operation : undefined);
    }
  }

  private focusPaths = () => {
    const { focusPathsThroughVertex, vertexKey } = this.props;
    focusPathsThroughVertex(vertexKey);
  };

  private handleClick = () => {
    const { decorationValue, selectVertex, vertex } = this.props;
    if (decorationValue) selectVertex(vertex);
  };

  private hideVertex = () => {
    const { hideVertex, vertexKey } = this.props;
    hideVertex(vertexKey);
  };

  private setOperation = (operation: string) => {
    trackVertexSetOperation();
    this.props.setOperation(operation);
  };

  private updateChildren = () => {
    const { updateGenerationVisibility, vertexKey } = this.props;
    updateGenerationVisibility(vertexKey, EDirection.Downstream);
  };

  private updateParents = () => {
    const { updateGenerationVisibility, vertexKey } = this.props;
    updateGenerationVisibility(vertexKey, EDirection.Upstream);
  };

  private viewTraces = () => {
    trackViewTraces();
    const { vertexKey, getVisiblePathElems } = this.props;
    const elems = getVisiblePathElems(vertexKey);
    if (elems) {
      const urlIds: Set<string> = new Set();
      let currLength = MIN_LENGTH;
      // Because there is a limit on traceIDs, attempt to get some from each elem rather than all from one.
      const allIDs = elems.map(({ memberOf }) => memberOf.traceIDs.slice());
      while (allIDs.length) {
        const ids = allIDs.shift();
        if (ids && ids.length) {
          const id = ids.pop();
          if (id && !urlIds.has(id)) {
            // Keep track of the length, then break if it is too long, to avoid opening a tab with a URL that
            // the backend cannot process, even if there are more traceIDs
            currLength += PARAM_NAME_LENGTH + id.length;
            if (currLength > MAX_LENGTH) {
              break;
            }
            urlIds.add(id);
            if (urlIds.size >= MAX_LINKED_TRACES) {
              break;
            }
          }
          allIDs.push(ids);
        }
      }
      window.open(getSearchUrl({ traceID: Array.from(urlIds) }), '_blank');
    }
  };

  private onMouseUx = (event: React.MouseEvent<HTMLElement>) => {
    const { getGenerationVisibility, getVisiblePathElems, setViewModifier, vertexKey } = this.props;
    const hovered = event.type === 'mouseover';
    const visIndices = hovered
      ? (getVisiblePathElems(vertexKey) || []).map(({ visibilityIdx }) => {
          this.hoveredIndices.add(visibilityIdx);
          return visibilityIdx;
        })
      : Array.from(this.hoveredIndices);
    setViewModifier(visIndices, EViewModifier.Hovered, hovered);

    if (hovered) {
      this.setState({
        childrenVisibility: getGenerationVisibility(vertexKey, EDirection.Downstream),
        parentVisibility: getGenerationVisibility(vertexKey, EDirection.Upstream),
      });
    } else this.hoveredIndices.clear();
  };

  render() {
    const { childrenVisibility, parentVisibility } = this.state;
    const {
      decorationProgressbar,
      decorationValue,
      focalNodeUrl,
      isFocalNode,
      isPositioned,
      operation,
      service,
    } = this.props;

    const { radius, svcWidth, opWidth, svcMarginTop } = calcPositioning(service, operation);
    const trueRadius = decorationProgressbar ? RADIUS - PROGRESS_BAR_STROKE_WIDTH : RADIUS;
    const scaleFactor = trueRadius / radius;
    const transform = `translate(${RADIUS - radius}px, ${RADIUS - radius}px) scale(${scaleFactor})`;

    return (
      <div className="DdgNodeContent" onMouseOver={this.onMouseUx} onMouseOut={this.onMouseUx}>
        {decorationProgressbar}
        <div
          className={cx('DdgNodeContent--core', {
            'is-decorated': decorationValue,
            'is-focalNode': isFocalNode,
            'is-missingDecoration': typeof decorationValue === 'string',
            'is-positioned': isPositioned,
          })}
          onClick={this.handleClick}
          role="button"
          style={{ width: `${radius * 2}px`, height: `${radius * 2}px`, transform }}
        >
          <div className="DdgNodeContent--labelWrapper">
            <h4
              className="DdgNodeContent--label"
              style={{ marginTop: `${svcMarginTop}px`, width: `${svcWidth}px` }}
            >
              <BreakableText text={service} wordRegexp={WORD_RX} />
            </h4>
            {operation && (
              <div
                className="DdgNodeContent--label"
                style={{ paddingTop: `${OP_PADDING_TOP}px`, width: `${opWidth}px` }}
              >
                {Array.isArray(operation) ? (
                  <Popover
                    content={<FilteredList options={operation} value={null} setValue={this.setOperation} />}
                    placement="bottom"
                    title="Select Operation to Filter Graph"
                  >
                    <span>{`${operation.length} Operations`}</span>
                  </Popover>
                ) : (
                  <BreakableText text={operation} wordRegexp={WORD_RX} />
                )}
              </div>
            )}
          </div>
        </div>
        <div className="DdgNodeContent--actionsWrapper">
          {focalNodeUrl && (
            <a href={focalNodeUrl} className="DdgNodeContent--actionsItem" onClick={trackSetFocus}>
              <span className="DdgNodeContent--actionsItemIconWrapper">{setFocusIcon}</span>
              <span className="DdgNodeContent--actionsItemText">Set focus</span>
            </a>
          )}
          <a className="DdgNodeContent--actionsItem" onClick={this.viewTraces} role="button">
            <span className="DdgNodeContent--actionsItemIconWrapper">
              <NewWindowIcon />
            </span>
            <span className="DdgNodeContent--actionsItemText">View traces</span>
          </a>
          {!isFocalNode && (
            <a className="DdgNodeContent--actionsItem" onClick={this.focusPaths} role="button">
              <span className="DdgNodeContent--actionsItemIconWrapper">
                <IoAndroidLocate />
              </span>
              <span className="DdgNodeContent--actionsItemText">Focus paths through this node</span>
            </a>
          )}
          {!isFocalNode && (
            <a className="DdgNodeContent--actionsItem" onClick={this.hideVertex} role="button">
              <span className="DdgNodeContent--actionsItemIconWrapper">
                <MdVisibilityOff />
              </span>
              <span className="DdgNodeContent--actionsItemText">Hide node</span>
            </a>
          )}
          {parentVisibility && (
            <a className="DdgNodeContent--actionsItem" onClick={this.updateParents} role="button">
              <span className="DdgNodeContent--actionsItemIconWrapper">
                <Checkbox
                  checked={parentVisibility === ECheckedStatus.Full}
                  indeterminate={parentVisibility === ECheckedStatus.Partial}
                />
              </span>
              <span className="DdgNodeContent--actionsItemText">View Parents</span>
            </a>
          )}
          {childrenVisibility && (
            <a className="DdgNodeContent--actionsItem" onClick={this.updateChildren} role="button">
              <span className="DdgNodeContent--actionsItemIconWrapper">
                <Checkbox
                  checked={childrenVisibility === ECheckedStatus.Full}
                  indeterminate={childrenVisibility === ECheckedStatus.Partial}
                />
              </span>
              <span className="DdgNodeContent--actionsItemText">View Children</span>
            </a>
          )}
        </div>
      </div>
    );
  }
}

export function mapDispatchToProps(dispatch: Dispatch<ReduxState>): TDispatchProps {
  const { getDecoration } = bindActionCreators(padActions, dispatch);

  return {
    getDecoration,
  };
}

const DdgNodeContent = connect(
  extractDecorationFromState,
  mapDispatchToProps
)(UnconnectedDdgNodeContent);

export default DdgNodeContent;
