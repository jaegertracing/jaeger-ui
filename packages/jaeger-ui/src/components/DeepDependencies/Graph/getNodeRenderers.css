/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DdgNode--VectorBorder {
  fill: #fff;
  stroke: #bbb;
  stroke-width: 2;
}

.DdgNode--VectorBorder.is-findMatch {
  /* This color may be overridden or removed when other coloring rationales are added */
  stroke: #fadb14;
}

.DdgNode--VectorBorder.is-focalNode {
  stroke: #eb2f96;
}

.DdgNode--VectorBorder.is-pathHovered,
.DdgNode--VectorBorder.is-hovered {
  stroke-opacity: 1;
}

.DdgNode--HtmlEmphasis {
  border-radius: 100%;
  bottom: -20px;
  content: '';
  left: -20px;
  position: absolute;
  right: -20px;
  top: -20px;
  z-index: -1;
  margin: 1px 0 0 1px;
}

.DdgNode--HtmlEmphasis.is-findMatch {
  background-image: repeating-linear-gradient(-60deg, #fffb8f, #fff566 3px, #fffb8f 5px);
  border: 1px dashed rgba(173, 139, 0, 0.5);
}

.DdgNode--HtmlEmphasis.is-focalNode {
  background-image: repeating-linear-gradient(30deg, transparent, #eb2f96 3px, transparent 5px);
  border: 1px dashed #eb2f96;
}

.DdgNode--HtmlEmphasis.is-focalNode.is-findMatch {
  background-image: repeating-linear-gradient(30deg, #fffb8f, #eb2f96 3px, #fffb8f 5px);
}

.DdgNode--VectorFindEmphasis--colorBand {
  stroke: #fef876;
  stroke-width: 10;
}
