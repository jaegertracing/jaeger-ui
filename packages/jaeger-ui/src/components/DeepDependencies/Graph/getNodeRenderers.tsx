// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';
import { TLayoutVertex } from '@jaegertracing/plexus/lib/types';

import { TDdgVertex, EViewModifier } from '../../../model/ddg/types';

import './getNodeRenderers.css';

export default function getNodeRenderers(findMatches: Set<string>, viewModifiers: Map<string, number>) {
  function vectorBorder(lv: TLayoutVertex<TDdgVertex>) {
    // eslint-disable-next-line no-bitwise
    const isHovered = (viewModifiers.get(lv.vertex.key) || 0) & EViewModifier.Hovered;
    // eslint-disable-next-line no-bitwise
    const isPathHovered = (viewModifiers.get(lv.vertex.key) || 0) & EViewModifier.PathHovered;
    const className = cx('DdgNode--VectorBorder', {
      'is-findMatch': findMatches.has(lv.vertex.key),
      'is-hovered': isHovered,
      'is-pathHovered': isPathHovered,
      'is-focalNode': lv.vertex.isFocalNode,
    });
    return (
      <circle
        className={className}
        vectorEffect="non-scaling-stroke"
        r={lv.width / 2 - 1}
        cx={lv.width / 2}
        cy={lv.width / 2}
      />
    );
  }

  function htmlEmphasis(lv: TLayoutVertex<{ isFocalNode: boolean }>) {
    const matchClasses = cx({
      'is-findMatch': findMatches.has(lv.vertex.key),
      'is-focalNode': lv.vertex.isFocalNode,
    });
    if (!matchClasses) {
      return null;
    }
    return <div className={`DdgNode--HtmlEmphasis ${matchClasses}`} />;
  }

  if (!findMatches.size) {
    return {
      vectorBorder,
      htmlEmphasis,
      vectorFindColorBand: null,
    };
  }

  function vectorFindColorBand(lv: TLayoutVertex) {
    if (!findMatches.has(lv.vertex.key)) {
      return null;
    }

    return (
      <circle
        className="DdgNode--VectorFindEmphasis--colorBand"
        vectorEffect="non-scaling-stroke"
        r={lv.width / 2 - 1}
        cx={lv.width / 2}
        cy={lv.width / 2}
      />
    );
  }

  return {
    htmlEmphasis,
    vectorBorder,
    vectorFindColorBand,
  };
}
