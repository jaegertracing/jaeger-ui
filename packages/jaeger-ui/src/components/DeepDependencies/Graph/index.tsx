// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import memoize from 'memoize-one';
import { Digraph, LayoutManager } from '@jaegertracing/plexus';
import { TSetProps, TFromGraphStateFn, TDefEntry } from '@jaegertracing/plexus/lib/Digraph/types';
import { TEdge } from '@jaegertracing/plexus/lib/types';
import TNonEmptyArray from '@jaegertracing/plexus/lib/types/TNonEmptyArray';

import { getNodeRenderer, measureNode } from './DdgNodeContent';
import getNodeRenderers from './getNodeRenderers';
import getSetOnEdge from './getSetOnEdge';
import {
  ECheckedStatus,
  EDdgDensity,
  EDirection,
  EViewModifier,
  PathElem,
  TDdgVertex,
} from '../../../model/ddg/types';

import './index.css';

type TProps = {
  baseUrl: string;
  density: EDdgDensity;
  edges: TEdge[];
  edgesViewModifiers: Map<string, number>;
  extraUrlArgs?: { [key: string]: unknown };
  focusPathsThroughVertex: (vertexKey: string) => void;
  getGenerationVisibility: (vertexKey: string, direction: EDirection) => ECheckedStatus | null;
  getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
  hideVertex: (vertexKey: string) => void;
  selectVertex: (selectedVertex: TDdgVertex) => void;
  setOperation: (operation: string) => void;
  setViewModifier: (visIndices: number[], viewModifier: EViewModifier, enable: boolean) => void;
  uiFindMatches: Set<string> | undefined;
  updateGenerationVisibility: (vertexKey: string, direction: EDirection) => void;
  vertices: TDdgVertex[];
  verticesViewModifiers: Map<string, number>;
};

// exported for tests
// The dichotomy between w/ & w/o VMs assumes that any edge VM neccesitates unmodified edges are de-emphasized
export const setOnEdgesContainer: Record<string, TSetProps<TFromGraphStateFn<unknown, unknown>>> = {
  withViewModifiers: [{ className: 'Ddg--Edges is-withViewModifiers' }],
  withoutViewModifiers: [Digraph.propsFactories.scaleStrokeOpacityStrongest, { className: 'Ddg--Edges' }],
};

// exported for tests
// The dichotomy between w/ & w/o VMs assumes that any vertex VM makes unmodified vertices de-emphasized
export const setOnVectorBorderContainerWithViewModifiers: TSetProps<
  TFromGraphStateFn<TDdgVertex, unknown>
> = {
  className: 'DdgVectorBorders is-withViewModifiers',
};

const edgesDefs: TNonEmptyArray<TDefEntry<TDdgVertex, unknown>> = [
  { localId: 'arrow' },
  { localId: 'arrow-hovered', setOnEntry: { className: 'Ddg--Arrow is-pathHovered' } },
];

export default class Graph extends PureComponent<TProps> {
  private getNodeRenderers = memoize(getNodeRenderers);
  private getNodeContentRenderer = memoize(getNodeRenderer);
  private getSetOnEdge = memoize(getSetOnEdge);

  private layoutManager: LayoutManager = new LayoutManager({
    nodesep: 0.55,
    ranksep: 1.5,
    rankdir: 'TB',
    shape: 'circle',
    splines: 'polyline',
    useDotEdges: true,
  });

  private emptyFindSet: Set<string> = new Set();

  componentWillUnmount() {
    this.layoutManager.stopAndRelease();
  }

  render() {
    const {
      baseUrl,
      density,
      edges,
      edgesViewModifiers,
      extraUrlArgs,
      focusPathsThroughVertex,
      getGenerationVisibility,
      getVisiblePathElems,
      hideVertex,
      selectVertex,
      setOperation,
      setViewModifier,
      uiFindMatches,
      updateGenerationVisibility,
      vertices,
      verticesViewModifiers,
    } = this.props;
    const nodeRenderers = this.getNodeRenderers(uiFindMatches || this.emptyFindSet, verticesViewModifiers);

    return (
      <Digraph<TDdgVertex>
        minimap
        zoom
        minimapClassName="u-miniMap"
        layoutManager={this.layoutManager}
        edges={edges}
        vertices={vertices}
        measurableNodesKey="nodes/content"
        layers={[
          {
            key: 'nodes/find-emphasis/vector-color-band',
            layerType: 'svg',
            renderNode: nodeRenderers.vectorFindColorBand,
          },
          {
            key: 'nodes/find-emphasis/html',
            layerType: 'html',
            renderNode: nodeRenderers.htmlEmphasis,
          },
          {
            key: 'nodes/vector-border',
            layerType: 'svg',
            renderNode: nodeRenderers.vectorBorder,
            setOnContainer: verticesViewModifiers.size
              ? setOnVectorBorderContainerWithViewModifiers
              : Digraph.propsFactories.scaleStrokeOpacityStrongest,
          },
          {
            key: 'edges',
            layerType: 'svg',
            edges: true,
            defs: edgesDefs,
            markerEndId: 'arrow',
            setOnContainer: edgesViewModifiers.size
              ? setOnEdgesContainer.withViewModifiers
              : setOnEdgesContainer.withoutViewModifiers,
            setOnEdge: this.getSetOnEdge(edgesViewModifiers),
          },
          {
            key: 'nodes/content',
            layerType: 'html',
            measurable: true,
            measureNode,
            renderNode: this.getNodeContentRenderer({
              baseUrl,
              density,
              extraUrlArgs,
              focusPathsThroughVertex,
              getGenerationVisibility,
              getVisiblePathElems,
              hideVertex,
              selectVertex,
              setOperation,
              setViewModifier,
              updateGenerationVisibility,
            }),
          },
        ]}
      />
    );
  }
}
