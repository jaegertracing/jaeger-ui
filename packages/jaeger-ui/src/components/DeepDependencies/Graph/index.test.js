// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';
import { Digraph, LayoutManager } from '@jaegertracing/plexus';

import Graph, { setOnEdgesContainer, setOnVectorBorderContainerWithViewModifiers } from './index';

import { EViewModifier } from '../../../model/ddg/types';

describe('<Graph />', () => {
  const vertices = [...new Array(10)].map((_, i) => ({ key: `key${i}` }));
  const edges = [
    {
      from: vertices[0].key,
      to: vertices[1].key,
    },
    {
      from: vertices[1].key,
      to: vertices[2].key,
    },
  ];

  const props = {
    edges,
    edgesViewModifiers: new Map(),
    vertices,
    verticesViewModifiers: new Map(),
  };

  describe('constructor', () => {
    it('creates layout manager', () => {
      const graph = new Graph(props);
      expect(graph.layoutManager instanceof LayoutManager).toBe(true);
    });
  });

  describe('render', () => {
    let wrapper;
    let plexusGraph;

    beforeEach(() => {
      wrapper = shallow(<Graph {...props} />);
      plexusGraph = wrapper.find(Digraph);
    });

    it('renders provided edges and vertices', () => {
      expect(plexusGraph.prop('edges')).toEqual(edges);
      expect(plexusGraph.prop('vertices')).toEqual(vertices);
      expect(wrapper).toMatchSnapshot();
    });

    it('de-emphasizes non-matching edges iff edgeVMs are present', () => {
      expect(plexusGraph.prop('layers')[3].setOnContainer).toBe(setOnEdgesContainer.withoutViewModifiers);

      wrapper.setProps({ edgesViewModifiers: new Map([[0, EViewModifier.Emphasized]]) });
      plexusGraph = wrapper.find(Digraph);
      expect(plexusGraph.prop('layers')[3].setOnContainer).toBe(setOnEdgesContainer.withViewModifiers);
    });

    it('de-emphasizes non-matching vertices iff vertexVMs are present', () => {
      expect(plexusGraph.prop('layers')[2].setOnContainer).toBe(
        Digraph.propsFactories.scaleStrokeOpacityStrongest
      );

      wrapper.setProps({ verticesViewModifiers: new Map([[0, EViewModifier.Emphasized]]) });
      plexusGraph = wrapper.find(Digraph);
      expect(plexusGraph.prop('layers')[2].setOnContainer).toBe(setOnVectorBorderContainerWithViewModifiers);
    });
  });

  describe('clean up', () => {
    it('stops LayoutManager before unmounting', () => {
      const wrapper = shallow(<Graph {...props} />);
      const stopAndReleaseSpy = jest.spyOn(wrapper.instance().layoutManager, 'stopAndRelease');
      wrapper.unmount();
      expect(stopAndReleaseSpy).toHaveBeenCalledTimes(1);
    });
  });
});
