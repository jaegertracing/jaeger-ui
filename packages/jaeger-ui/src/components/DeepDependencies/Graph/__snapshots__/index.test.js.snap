// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Graph /> render renders provided edges and vertices 1`] = `
<Digraph
  className=""
  classNamePrefix="plexus"
  edges={
    Array [
      Object {
        "from": "key0",
        "to": "key1",
      },
      Object {
        "from": "key1",
        "to": "key2",
      },
    ]
  }
  layers={
    Array [
      Object {
        "key": "nodes/find-emphasis/vector-color-band",
        "layerType": "svg",
        "renderNode": null,
      },
      Object {
        "key": "nodes/find-emphasis/html",
        "layerType": "html",
        "renderNode": [Function],
      },
      Object {
        "key": "nodes/vector-border",
        "layerType": "svg",
        "renderNode": [Function],
        "setOnContainer": [Function],
      },
      Object {
        "defs": Array [
          Object {
            "localId": "arrow",
          },
          Object {
            "localId": "arrow-hovered",
            "setOnEntry": Object {
              "className": "Ddg--Arrow is-pathHovered",
            },
          },
        ],
        "edges": true,
        "key": "edges",
        "layerType": "svg",
        "markerEndId": "arrow",
        "setOnContainer": Array [
          [Function],
          Object {
            "className": "Ddg--Edges",
          },
        ],
        "setOnEdge": Object {
          "className": "Ddg--Edge",
        },
      },
      Object {
        "key": "nodes/content",
        "layerType": "html",
        "measurable": true,
        "measureNode": [Function],
        "renderNode": [Function],
      },
    ]
  }
  layoutManager={
    LayoutManager {
      "_handleUpdate": [Function],
      "coordinator": Coordinator {
        "_handleVizWorkerError": [Function],
        "_handleVizWorkerMessage": [Function],
        "_handleVizWorkerMessageError": [Function],
        "busyWorkers": Array [],
        "callback": [Function],
        "currentLayout": null,
        "idleWorkers": Array [],
        "nextWorkerId": 0,
      },
      "layoutId": 0,
      "options": Object {
        "nodesep": 0.55,
        "rankdir": "TB",
        "ranksep": 1.5,
        "shape": "circle",
        "splines": "polyline",
        "useDotEdges": true,
      },
      "pendingResult": null,
    }
  }
  measurableNodesKey="nodes/content"
  minimap={true}
  minimapClassName="u-miniMap"
  vertices={
    Array [
      Object {
        "key": "key0",
      },
      Object {
        "key": "key1",
      },
      Object {
        "key": "key2",
      },
      Object {
        "key": "key3",
      },
      Object {
        "key": "key4",
      },
      Object {
        "key": "key5",
      },
      Object {
        "key": "key6",
      },
      Object {
        "key": "key7",
      },
      Object {
        "key": "key8",
      },
      Object {
        "key": "key9",
      },
    ]
  }
  zoom={true}
/>
`;
